packages/dx-react-grid/docs/guides/custom-plugin-development.md

```diff
@@ -0,0 +1,702 @@
+    const newFilterExpression = { ...filterExpression || { filters: [], operator: "and" } };
+    if (toolbarFilterValue) {
+      newFilterExpression.filters.push({
+        columnName: toolbarFilterColumnName,
+        value: toolbarFilterValue
+      });
+    }
+    return newFilterExpression;
+}
+
+export class ToolbarFilterState extends React.PureComponent {
+  ...
+  render() {
+    ...
+    return (
+      <Plugin
+        name="ToolbarFilterState"
+      >
+        ...
+        <Getter name="filterExpression" computed={filterExpressionComputed} />
+      </Plugin>
+    );
+  }
+}
+```
+
+Now our grid can filter items by the `CompanyName` column.
+
+
+
+## Final Steps
+
+Let's complete this task by displaying the current column caption near the select box and adding the `Clear` button for clearing the selected value. For this task we will perform the following steps:
+
+* Declare the `clearToolbarFilterValue` action in the `ToolbarFilterState` plugin and clear the `filterValue` state property in the action.
+* Add the `Button` widget near the select box in the `ToolbarFilter` plugin and use the `clearToolbarFilterValue` action as the `onClick` event handler.
+* Get the `toolbarFilterColumnName` getter value in the `ToolbarFilterState` plugin.
+* Since we want to get a caption of the `CompanyName` column, we will need to find the required column by its name and get the caption. For this task we will get the `columns` getter value that is exported by the [Grid](../reference/grid.md) plugin.
+
+The following code snippets show this approach in action:
+*file toolbar-filter-state.js:*
+```jsx
+...
+
+export class ToolbarFilterState extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    ...
+    this.columnName = this.props.defaultColumnName || null;
+    this.clearValue = this.clearValue.bind(this);
+  }
+
+  clearValue() {
+    this.setState({
+      filterValue: ""
+    });
+  }
+
+  render() {
+    ...
+    return (
+      <Plugin
+        name="ToolbarFilterState"
+      >
+        ...
+        <Action name="clearToolbarFilterValue" action={this.clearValue} />
+      </Plugin>
+    );
+  }
+}
+```
+
+*toolbar-filter.js:*
+```jsx
+...
+const getColumnTitle = (columns, toolbarFilterColumnName) => {
+  const column = columns.find(col => col.name === toolbarFilterColumnName);
+  return column ? column.title : "";
+}
+export class ToolbarFilter extends React.PureComponent {
+  render() {
+    return (
+      <Plugin
+        ...
+      >
+        <Template name="toolbarContent">
+          <TemplatePlaceholder />
+          <TemplateConnector>
+          {({
+              ...,
+              toolbarFilterColumnName,
+              columns
+            }, {
+              ...,
+              clearToolbarFilterValue
+            }) => (
+              <div>
+                <InputLabel htmlFor="filter-field">{getColumnTitle(columns, toolbarFilterColumnName)}:</InputLabel>
```

Comment:
There is no need to recalculate a column title each time filter is being rendered if the column name hasn't changed. It's enough to calculate it just one time when the column name has changed. So, it should be a computed exposed from the ToolbarFilterState plugin. I would combine it with columnName and export the filter column info (name + title).